cranelift/codegen/src/machinst/mod.rs

```diff
@@ -269,6 +272,48 @@ pub trait MachInstLabelUse: Clone + Copy + Debug + Eq {
     fn from_reloc(reloc: Reloc, addend: Addend) -> Option<Self>;
 }
 
+/// Classification of call instruction types for granular analysis.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum CallType {
+    /// Not a call instruction.
+    None,
+    /// Regular call that returns to the caller.
+    Regular,
+    /// Tail call that doesn't return to the caller.
+    TailCall,
+}
+
+/// Enhanced function type classification for optimization purposes.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum FunctionType {
+    /// Function makes no calls at all.
+    Leaf,
+    /// Function only makes tail calls.
+    TailCallOnly,
+    /// Function has regular calls (with or without tail calls).
+    NonLeaf,
+}
```

Comment:
A few naming thoughts:

- `FunctionType` evokes something more like a function signature to me; maybe `FunctionCalls` for this enum?
- Once we call it that, the arms might be a little clearer as `FunctionCalls::None`, `FunctionCalls::TailOnly`, and `FunctionCalls::Regular`.

Reply:
I don't mind FunctionCalls